graphql_compiler/query_pagination/query_parameterizer.py

```diff
@@ -241,12 +249,16 @@ def _add_pagination_filter_recursively(
                  the same operation removed.
         new_parameters: The parameters to use with the new_ast
     """
-    if not isinstance(node_ast, (FieldNode, InlineFragmentNode, OperationDefinitionNode)):
-        raise AssertionError(
-            f'Input AST is of type "{type(node_ast).__name__}", which should not be a selection.'
-        )
-
     if len(query_path) == 0:
+        if isinstance(node_ast, OperationDefinitionNode):
+            # N.B.: We cannot use assert_selection_is_a_field_or_inline_fragment_node() here, since
+            #       that function's return type annotation is a Union type that isn't compatible
+            #       with the generic type signature of _add_pagination_filter_at_node().
```

Comment:
Interesting how `assert_selection_is_a_field_or_inline_fragment_node` works with unions: "We used to know x is of type A, but we ran some assertions and now we're not sure if it's A or B"

Can't think of a cleaner solution.

Btw, we use `N.B.` and `NOTE` in the codebase. No need to standardize, but letting you know that `NOTE` gets highlighted by many text editors, just like `TODO` is.

Reply:
If `checked_cast()` works out, we might not need this note anymore. 🤞

Reply:
I wrote a `checked_cast_to_union2()` function but it turns out that even it can't do the trick with generics. I updated the note with my new observations, let me know if you want to talk about it.